File: Java/0023_merge_k_sorted_lists/0023_merge_k_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

public class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        Queue<ListNode> heap = new PriorityQueue<ListNode>(new Comparator<ListNode>() {
            @Override
            public int compare(ListNode o1, ListNode o2) {
                return o1.val - o2.val;
            }   
        });
        ListNode result = new ListNode(0);
        ListNode temp = result;
        for (ListNode listNode : lists) {
            if (listNode != null)
                heap.add(listNode);
        }   
        while (!heap.isEmpty()) {
            ListNode cur = heap.poll();
            temp.next = cur;
            if (cur.next != null) {
                heap.add(cur.next);
            }
            temp = temp.next;
        }         
        return result.next;
    }
}

File: Cplusplus/0023_merge_k_sorted_lists/0023_merge_k_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct Comparator {
    bool operator() (const ListNode* n1, const ListNode* n2) {
        return (n1->val > n2->val);
    }
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, Comparator> pq;         
        for (auto node : lists) {
            if (node)
                pq.push(node);
        } 
        ListNode* result = new ListNode(0);
        ListNode* temp = result;  
        while (!pq.empty()) {
            ListNode* cur = pq.top();
            pq.pop();
            temp->next = cur;
            if (cur->next != NULL)
                pq.push(cur->next);
            temp = temp->next;
        }         
        return result->next;
    }
};
